rfcs/20201102-convert-verbose-into-benchdnn-problem/README.md

```diff
@@ -0,0 +1,50 @@
+# RFC: Convert DNNL_VERBOSE output into benchdnn problems
+
+## Motivation
+
+There are several requests from users and developers to enable DNNL_VERBOSE
+output be "runnable" from benchdnn. Such requests mostly consider the following
+scenario: there is some integration of oneDNN into some framework, and there is
+some topology to run. If validation shows performance or correctness regression,
+it is very hard and/or time consuming to narrow down the specific guilty layer,
+and those users usually send a log file with DNNL_VERBOSE output, which is
+really massive. Manual conversion of these outputs into benchdnn problems takes
+enormous amount of time, since the topology usually consists of 100+ layers. It
+is just inconvenient not to have some tool that allows to perform such
+conversion very fast.
+
+## Tool Requirements
+
+* The tool should accept either a single line or a text file with lines.
+* The output should be fully benchdnn-compatible.
+* The tool should be easy to use.
+
+## Proposal
+
+### Option 1 (recommended) - Python script.
+
+Use the power of Python to create such tool. It will take the input and produce
+benchdnn batch files separated by drivers. Script advantages:
+* Flexibility in terms of features support.
+* Not working script will not affect a release.
+* It may have compatibility options between different options in benchdnn
+  drivers inside and will not depend on a specific version of the library and
+  benchdnn.
+
+### Option 2 - native benchdnn support.
+
+Implement such support as a standalone benchdnn driver. All pros of a Python
+script seems to be a cons of C++ solution:
+* Each feature should be tested through regular oneDNN validation cycles,
+  since it's a main validation tool.
+* Errors and bugs in code will likely lead to patch fixes for a release.
+* Versinong - earlier versions may lack of some suitable features.
+
+## Testing
+
+There is no clear vision about how the testing should like in either case. The
```

Comment:
1. Take benchdnn input file
2. Run benchdnn on this input file with DNNL_VERBOSE=1
3. Feed the output log to the script
4. Compare the output with original input file (might need to have an input file in "canonical" form)
5. ...
6. Profit?

Reply:
@emfomenk I wrote the same comment, but deleted it ... The problem is that you have to execute primitives which will consume additional and unnecessary time in oneDNN testing.

Reply:
List mode?

Reply:
Sorry, it's not gonna work.

Reply:
But anyway, we can run several cases just to make sure that it works.

Reply:
Unlike real oneDNN testing, I think here it is enough to run several problems for each driver. Morever, it is fine to use tiny sizes, so the testing won't take much time. It is also oK to run in on Linux only.

But I agree that this approach might be rather complicated, and maybe it is just fine to have a "validation input file" of the following form:

``` bash
$ ls
verbose2benchdnn.py
validate.txt

$ cat validate.txt
<dnnl-verbose-line-1>
<expect-benchdnn-input-line-1>

<dnnl-verbose-line-2>
<expect-benchdnn-input-line-2>

<dnnl-verbose-line-3>
<expect-benchdnn-input-line-3>

...

$ python verbose2benchdnn.py --validate validate.txt; echo $?
0
```

Reply:
Thanks for ideas, this approach will very likely work. I've re-worded the section about testing. It mostly describes your ideas.